### ISSUE TITLE
Dart Error: Dart_LookupLibrary When building/running app in release mode.

### ISSUE THREAD

**Issue by atharvsharma1998** [radarlabs/flutter-radar#41]
I have been facing an issue related to working of radar in release mode. While in debug mode the library is working fine with no errors, I think it is being caused due to obfuscation due to which R8 is removing the native code/methods [https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/aot/entry_point_pragma.md]. I tried adding ```@pragma("vm:entry-point")``` to the flutter_radar.dart but it didn't solve the issue. This same issue is reproducible in the radar example code. Help from the team would be great.

flutter doctor:
```
[✓] Flutter (Channel stable, 3.19.3, on macOS 13.5 22G74 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] VS Code (version 1.87.2)
[✓] Connected device (3 available)
[✓] Network resources
• No issues found!
```
LOGS:
```
E/flutter ( 5294): [ERROR:flutter/shell/common/shell.cc(117)] Dart Error: Dart_LookupLibrary: library 'package:flutter_radar/flutter_radar.dart' not found.
E/flutter ( 5294): [ERROR:flutter/runtime/dart_isolate.cc(677)] Could not resolve main entrypoint function.
E/flutter ( 5294): [ERROR:flutter/runtime/dart_isolate.cc(168)] Could not run the run main Dart entrypoint.
E/flutter ( 5294): [ERROR:flutter/runtime/runtime_controller.cc(462)] Could not create root isolate.
E/flutter ( 5294): [ERROR:flutter/shell/common/shell.cc(669)] Could not launch engine with configuration.
```

Im not sure it will help in explanation of this issue or debugging as this issue is appearing in stock radar example too. But I have included proguard-rules in my personal project as its an requirement by flutter_local_notifcation package:
proguard-rules.pro
```
## Gson rules
# Gson uses generic type information stored in a class file when working with fields. Proguard
# removes such information by default, so configure it to keep all of it.
#-keepattributes InnerClasses
-keepattributes Signature

# For using GSON @Expose annotation
-keepattributes *Annotation*

# Gson specific classes
-dontwarn sun.misc.**
#-keep class com.google.gson.stream.** { *; }

# Prevent proguard from stripping interface information from TypeAdapter, TypeAdapterFactory,
# JsonSerializer, JsonDeserializer instances (so they can be used in @JsonAdapter)
-keep class * extends com.google.gson.TypeAdapter
-keep class * implements com.google.gson.TypeAdapterFactory
-keep class * implements com.google.gson.JsonSerializer
-keep class * implements com.google.gson.JsonDeserializer

# Prevent R8 from leaving Data object members always null
-keepclassmembers,allowobfuscation class * {
  @com.google.gson.annotations.SerializedName <fields>;
}

# Retain generic signatures of TypeToken and its subclasses with R8 version 3.0 and higher.
-keep,allowobfuscation,allowshrinking class com.google.gson.reflect.TypeToken
-keep,allowobfuscation,allowshrinking class * extends com.google.gson.reflect.TypeToken
```

Edit: adding @pragma("vm:entry-point") before class Radar make the error E/flutter ( 6472): [ERROR:flutter/shell/common/shell.cc(117)] Dart Error: Dart_LookupLibrary: library 'package:flutter_radar/flutter_radar.dart' not found. go away. 

```
@pragma("vm:entry-point")
class Radar {
  static const MethodChannel _channel = const MethodChannel('flutter_radar');
```

But the library still doesn't works.

**Comment by KennyHuRadar** [comment]
can you try adding the vm entry point to the `callbackDispatcher` instead like such

```
@pragma('vm:entry-point')
void callbackDispatcher() {
  const MethodChannel _backgroundChannel =
      MethodChannel('flutter_radar_background');
  WidgetsFlutterBinding.ensureInitialized();

  _backgroundChannel.setMethodCallHandler((MethodCall call) async {
    final args = call.arguments;
    final CallbackHandle handle = CallbackHandle.fromRawHandle(args[0]);
    final Function? callback = PluginUtilities.getCallbackFromHandle(handle);
    final Map res = args[1];
    
    callback?.call(res);
  });
}
```

It may also be helpful to add the following to your proguard-rules.pro while debugging:
```
// this will force R8 to not keep all Radar related classes
-keep class io.radar.** { *; }
// This will print all the classes that were removed by the obfuscation
-printusage debug-dir/usage.txt
// This will print the classes that "seed" what is kept by the obfuscation
-printseeds debug-dir/seeds.txt
```

**Comment by atharvsharma1998** [comment]
Thanks for the fix, but using it didn't work. I'm getting error
```
E/flutter (26098): [ERROR:flutter/lib/ui/dart_runtime_hooks.cc(38)] Dart Error: Dart_LookupLibrary: library 'package:hot_frontend/screens/MapScreen.dart' not found.

```

**Comment by atharvsharma1998** [comment]
I haven't been able to exactly follow the workaround as the file structure/logic have changed so couldn't tell if this solution works or not. Also, is this issue reproducible by radar team as I'm getting it in example radar project too?
https://github.com/flutter/flutter/issues/91841#issuecomment-1006026551
https://github.com/flutter/flutter/issues/91841#issuecomment-1027286001

**Comment by KennyHuRadar** [comment]
I am seeing a different error on the example app. 
```
E/flutter (27400): [ERROR:flutter/lib/ui/dart_runtime_hooks.cc(38)] Dart Error: Class '_MyAppState' not found in library ''.
E/flutter (27400): [ERROR:flutter/shell/common/shell.cc(117)] Dart Error: Class '_MyAppState' not found in library ''.
```
I think the new entry point annotation was able to resolve the issue of the obfuscation but may introduce a new issue. We are looking into it.

**Comment by KennyHuRadar** [comment]
I was able to suppress the errors that I was getting on the example app by adding the same annotation to the static methods we want to use as callbacks.

```
@pragma('vm:entry-point')
static void onEvents(Map res) {
print('📍📍 onEvents: $res');
}
  .....
Radar.onEvents(onEvents);
```

This is done in addition to adding the annotation to the `callbackDispatcher` as described in https://github.com/radarlabs/flutter-radar/issues/41#issuecomment-2045927456

**Comment by atharvsharma1998** [comment]
yes adding ```@pragma('vm:entry-point')``` to callbackDispatcher and the onEvents method solved the issue. Thanks. is it recommended to be added to every event listener?

**Comment by KennyHuRadar** [comment]
Yes, all event listeners are "passed into the native layer". As such, we need to add the annotation for them so that they can be executed when the native code executes the callback.

We will be cutting a release with this fix soon! Hopefully this workaround is sufficient in the meantime.

**Comment by atharvsharma1998** [comment]
issue patched in v3.9.1
